llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp

```diff
@@ -1267,6 +1267,19 @@ Instruction *InstCombinerImpl::visitLShr(BinaryOperator &I) {
       match(Op1, m_SpecificIntAllowUndef(BitWidth - 1)))
     return new ZExtInst(Builder.CreateIsNotNeg(X, "isnotneg"), Ty);
 
+  // Special Case:
+  // if both the add and the shift are nuw, we can omit the AND entirely
+  // ((X << Y) nuw + Z nuw) >>u Z --> (X + (Y >>u Z))
+  Value *Y;
+  if (match(Op0, m_OneUse(m_c_NUWAdd((m_NUWShl(m_Value(X), m_Specific(Op1))),
+                                     m_Value(Y))))) {
+    Value *NewLshr = Builder.CreateLShr(Y, Op1, "", I.isExact());
```

Comment:
This exact is only valid if `I` is `exact`

Reply:
I know. Thats' why I.isExact is there

Reply:
Oh i misread. Sorry.

Can you please update the proof to include the non-exact version and non-nsw version.

Reply:
Sure: https://alive2.llvm.org/ce/z/wORNYm